docs: Improve testing tool docs (#3196)
confluentinc/ksql: docs/developer-guide/ksql-testing-tool.rst

      Test failed: Expected <900, {T_ID=90, NAME=ninety}> with timestamp=17000 but was <90, {T_ID=90, NAME=ninety}> with timestamp=17000

Query Execution in the KSQL Testing Tool
****************************************

> To effectively use the KSQL testing tool, you need to understand the query execution logic in the testing tool. Although the final results

Reviewer: ```suggestion
To use the KSQL testing tool effectively, you need to understand the query execution logic that it uses. Although the final results
```

> should be deterministic, the intermediate results in KSQL queries (Kafka Streams Apps) may vary based on several factors such as order

Reviewer: ```suggestion
should be deterministic, the intermediate results in KSQL queries (Kafka Streams Apps) may vary based on several factors, such as order
```

> of reading input or config properties such as producer buffer size. In order to make the composition of output for the test cases

Reviewer: ```suggestion
of reading input or config properties like the producer buffer size. To make the composition of output for the test cases
```

> simpler, the KSQL testing tool executes queries in a predictable way. Users should consider the following when preparing the output for the tests.

Reviewer: ```suggestion
simpler, the KSQL testing tool executes queries in a predictable way. Consider the following guidance when you prepare the output for your tests.
```

Input Consumption
-----------------

> The testing tool processes input messages for each query one by one and writes the generated message(s) for each input message into the

Reviewer: ```suggestion
The testing tool processes input messages for each query one-by-one and writes the generated message(s) for each input message into the
```

> result topic before processing the next input message. This means that for the queires running in the testing tool, we have the same behavior as

Reviewer: ```suggestion
result topic. This means that for the queries running in the testing tool, the behavior is the same as
```

>  when cache.max.bytes.buffering = 0. This is especially important in aggregate queries where we may not see some of the intermediate results in

Reviewer: ```suggestion
 when ``cache.max.bytes.buffering = 0``. This is especially important in aggregate queries where, you may not see some of the intermediate results in
```

>  real executions because of buffering, while in the testing tool execution we will create every possible intermediate results.

Reviewer: ```suggestion
 real executions because of buffering, while during execution  of the testing tool every possible intermediate result is created.
```

Kafka Cluster
-------------

> The KSQL testing tool does not use a real Kafka cluster and simulates the behavior of a cluster with single broker for the KSQL queries.

Reviewer: ```suggestion
The KSQL testing tool doesn't use a real Kafka cluster and simulates the behavior of a single-broker cluster for KSQL queries.
```

> This means that the testing tool ignores configurations such as number of partitions or replicas for the input and output topics.

Reviewer: ```suggestion
This means that the testing tool ignores configuration settings for the input and output topics, like the number of partitions or replicas.
```

Processing Order
----------------

> The testing tool processes the statements in the order that they are provided by the user. Therefore, for a given statement,

Reviewer: ```suggestion
The testing tool processes the statements in the order that you provide them. So for a given statement,
```

> only the statements before it can potentially affect it's results. This is a different behavior than KSQ cluster where statements that

Reviewer: ```suggestion
only the statements before it can potentially affect its results. This is a different behavior than what you see in an actual KSQL cluster, where statements that
```

> are submitted later can affect the output of a query. As an example, consider the following set of statements:

Reviewer: ```suggestion
are submitted later *can* affect the output of a query. for example, consider the following set of statements:
```

.. code:: sql

    CREATE STREAM orders (ORDERUNITS double) WITH (kafka_topic='test_topic', value_format='JSON');
    INSERT INTO orders VALUES(10.0);
    INSERT INTO orders VALUES(15.0);
    INSERT INTO orders VALUES(20.0);
    CREATE STREAM S1 AS SELECT ORDERUNITS, CASE WHEN orderunits < 2.0 THEN 'small' WHEN orderunits < 4.0 THEN 'medium' ELSE 'large' END AS case_resault FROM orders;
    INSERT INTO orders VALUES(25.0);
    INSERT INTO orders VALUES(30.0);

> If you run the above statements in a real KSQL cluster, you will see one result generated for each INSERT INTO statements (You will have 5 messages in the output).

Reviewer: ```suggestion
If you run the above statements in a real KSQL cluster, you see one result generated for each INSERT INTO statements, and you have five messages in the output.
```

> On the other hand, if you run the above statements in the testing tool, only the INSERT INTO statements before the CSAS query will generate results and the testing tool won't

Reviewer: ```suggestion
On the other hand, if you run the previous statements in the testing tool, only the INSERT INTO statements before the CSAS query generate results, and the testing tool won't
```

run the query for the messages generated by the INSERT INTO statements after the CSAS statement.

> Another important processing order to keep in mind for the testing tool is the order in which the input data for a query is processed. For a given query,

Reviewer: ```suggestion
Another important detail about processing order to keep in mind is the order in which the input data for a query is processed. For a given query,
```

> the testing tool will first process the input messages provided in the input file. After fully processing such messages

Reviewer: ```suggestion
the testing tool first processes the input messages provided in the input file. After fully processing these messages,
```

> the testing tool then inspects the source topics for the query in the simulated Kafka cluster and processes any messages in those topics.

Reviewer: ```suggestion
the testing tool inspects the source topics for the query in the simulated Kafka cluster and processes any messages in these topics.
```

> For the JOIN queries where we have more than one source topic, the testing tool first processes the left hand side topic and

Reviewer: ```suggestion
For the JOIN queries where we have more than one source topic, the testing tool first processes the left-side topic and
```

> then processes the right hand side topic.

Reviewer: ```suggestion
then processes the right-side topic.
```
